docs/reference/terminology.md

```diff
@@ -1,29 +1,944 @@
+
+**Gradient Accumulation Steps:**
+Number of micro-batches to accumulate before performing a parameter update.
+
+**Rematerialization (Remat) / Activation Checkpointing:**
+Recomputing activations in backward pass instead of storing them, trading compute for memory.
+
+**Remat Policy:**
+Strategy for which activations to save vs recompute. Options include (from fastest/highest HBM to slowest/lowest HBM): `full`, `save_out_proj`, `save_qkv_proj`, `save_dot_except_mlp`, `save_dot_except_mlpwi`, `save_dot_with_context_except_mlp`, `minimal`, `minimal_with_context`, `minimal_offloaded`, `qkv_proj_offloaded`, and `custom`.
+
+**Tiling / Chunking:**
+Partitioning large tensors into smaller blocks processed sequentially to reduce peak memory usage.
+
+**Vocabulary Tiling:**
+Tiling the final unembedding layer to avoid materializing the full logits tensor (vocab_size × sequence_length × batch_size), controlled by `num_vocab_tiling` config.
+
+**Offloading:**
+Moving optimizer state or parameters to host CPU memory to free accelerator HBM.
+
+**Optimizer Memory Host Offload:**
+Storing optimizer state (momentum, variance) on CPU memory instead of accelerator memory.
+
+**Parameter Memory Host Offload:**
+Storing model parameters on CPU memory, loading them to accelerator only when needed.
+
+**ZeRO-1 Sharding:**
+Sharding only optimizer state across devices, not weights or gradients.
+
+**ZeRO-3 Sharding:**
+Same as FSDP, sharding weights, gradients, and optimizer state across all devices.
+
+---
+
+## Checkpointing
+
+**Checkpoint:**
+Saved model state including parameters, optimizer state, and other metadata (step count, etc.).
+
+**Training Checkpoint:**
+Full checkpoint with both parameters and optimizer state, used to resume training.
+
+**Inference Checkpoint / Param-Only Checkpoint:**
+Checkpoint containing only model parameters without optimizer state, used for inference or fine-tuning.
+
+**Stacked Checkpoint / Scanned Format:**
+Checkpoint with parameters stacked for `jax.lax.scan`, providing memory-efficient loading.
+
+**Unstacked Checkpoint / Unscanned Format:**
+Parameters stored layer-by-layer, standard format for public models and cross-framework compatibility.
+
+**Orbax:**
+JAX checkpointing library used by MaxText, providing flexible and efficient checkpoint management. [Document](https://orbax.readthedocs.io/en/latest/index.html)
+
+**Async Checkpointing:**
+Non-blocking checkpoint saves allowing training to continue while checkpoint is being written.
+
+**Sync Checkpointing:**
+Blocking checkpoint saves that pause training until checkpoint is written.
+
+**Emergency Checkpoint:**
+Local checkpoint saved for fast recovery after interruption, typically stored on local disk.
+
+**Multi-Tier Checkpointing:**
+Experimental Orbax feature combining periodic GCS saves with local emergency checkpoints.
+
+**Checkpoint Conversion:**
+Transforming checkpoints between formats (e.g., HuggingFace PyTorch ↔ MaxText Orbax).
+
+**ArrayRecord:**
```

Comment:
I see this is also in the Data pipeline section, which seems to be more related. Should we remove here?